CatalogService.API/EndpointDefinitions/ItemsEndpointDefinition.cs

```diff
@@ -0,0 +1,99 @@
+﻿using CatalogService.API.Core;
+using CatalogService.API.Models;
+using CatalogService.API.Utils;
+using CatalogService.Core.Entities;
+using CatalogService.Core.Interfaces;
+using CatalogService.Core.Models;
+using CatalogService.Core.Validators;
+using CatalogService.Infrastructure.Repositories;
+using Dapper;
+using FluentValidation;
+using Microsoft.AspNetCore.Mvc;
+
+namespace CatalogService.API.EndpointDefinitions;
+
+public class ItemsEndpointDefinition : IEndpointDefinition
+{
+    public void DefineEndpoints(WebApplication app)
+    {
+        app.MapGet("/items", ListItems);
+        app.MapPost("/items", AddItem);
+        app.MapPut("/items/{id}", UpdateItem);
+        app.MapDelete("/items/{id}", DeleteItem);
+    }
+
+    public void DefineServices(IServiceCollection services)
+    {
+        services.AddSingleton<IItemRepository, ItemRepository>();
+        services.AddSingleton<IValidator<ItemModel>, ItemValidator>();
+    }
+
+    private static async Task<IResult> ListItems([FromServices] IItemRepository repository,
+        [FromServices] IValidator<ItemModel> validator,
+        [FromQuery(Name = "categoryId")] string? categoryId = null,
+        [FromQuery(Name = "page")] ushort page = 1,
+        [FromQuery(Name = "pageSize")] ushort pageSize = 20,
+        CancellationToken cancellationToken = default)
+    {
+        var paginationValidation = PaginationValidation.ValidatePaginationFields(page, pageSize);
+
+        if (paginationValidation.IsT1)
+            return Results.BadRequest(paginationValidation.AsT1);
+
+        var item = new Item(validator, repository, cancellationToken);
+        var response = await item.GetAll(categoryId, --page, pageSize);
+
+        return response.Match(
+            list => Results.Ok(list.Select( i => new ItemDetailResponse(i))),
+            _ => Results.Ok(
+                Array.Empty<ItemModel>()));
+    }
+
+    private static async Task<IResult> AddItem([FromBody] AddItemModel model,
+        [FromServices] IItemRepository repository,
+        [FromServices] IValidator<ItemModel> validator,

```

Comment:
It is certainly a way to implement it, could you provide me some benefits of this implementation in comparison to constructor injection?

Reply:
The main benefit with this kind of implementation is regarding the performance against a traditional WebAPI with controllers. Here that's clearly not the case as the performance improvement is seen on nanoseconds, but it's something to keep in mind in my opinion.